### PR TITLE
Avoid redundant warnings from explicit cast

### LINKED ISSUE #30938: Duplicate ErrorCode.WRN_ConvertingNullableToNonNullable warning is reported at the same location
```
        [Fact]
        public void Test()
        {
            var source = @"
class Outer
{
    void M1(object x1, object? y1)
    {
        x1 = (object)y1;
    }
}
";

            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
                // (6,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x1 = (object)y1;
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(object)y1").WithLocation(6, 14),
                // (6,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
                //         x1 = (object)y1;
                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(object)y1").WithLocation(6, 14)
                );
        }
```

One is reported when we analyze conversion, another is reported when we analyze the assignment. Conversion preserves nullable state.

### PR DESCRIPTION
**Pull request by cston**
Fixes #30938.

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -3290,6 +3289,21 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
...
                 case ConversionKind.Identity:
+                    // If the operand is an explicit conversion, and this identity conversion
+                    // is converting to the same type including nullability, skip the conversion
+                    // to avoid reporting redundant warnings. Also check useLegacyWarnings
+                    // since that value was used when reporting warnings for the explicit cast.
+                    if (useLegacyWarnings && operandOpt?.Kind == BoundKind.Conversion)
+                    {
+                        var operandConversion = (BoundConversion)operandOpt;
+                        var explicitType = operandConversion.ConversionGroupOpt.ExplicitType;
+                        if (!explicitType.IsNull && explicitType.Equals(targetTypeWithNullability, TypeCompareKind.ConsiderEverything))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >ConsiderEverything [](start = 115, length = 18)

Should insignificant annotation differences be ignored for the purpose of this comparison?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Perhaps, although it's not clear how to get `NullableBasedOnAnalysis` or `NotNullableBasedOnAnalysis` in either the target type or the explicit cast type. I'll leave as is, and we can adjust if there are insignificant differences in the future.

**Comment by cston** [comment]
@jaredpar for approval.

**Pull request merged**